README.md
# Kafka Demo

Demonstrates writing to and processing simple timeseries data.

Producer generates a double sinusoid.

Python Consumer just echoes whatever is on the queue.

Clojure consumer calculates a first-order derivative and latest state.
Note that Clojure consumer needs to be ran after python script.

## Setup (Kafka)

Starto docker!
```bash
docker-compose up -d
```

Add kafka-demo topic inside docker
```
docker exec -it kafkademo_kafka_1 /bin/bash
kafka-topics.sh --create --bootstrap-server localhost:9092 --replication-factor 1 --partitions 1 --topic kafka-demo
```

## Setup (Python)

Install virtualenvwrapper, docker, docker-compose.

Create a virtual env:
```bash
mkvirtualenv --python=/usr/bin/python3 kafka-demo
```

Install requirements:
```bash
pip install --requirements requirements.txt
```

Run script and enjoy.
```bash
python demo.py
```

## Setup (Clojure)

Install leiningen.

Run project after setting up a topic
```bash
cd streamer
lein deps
lein run
```


streamer/CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. This change log follows the conventions of [keepachangelog.com](http://keepachangelog.com/).

## [Unreleased]
### Changed
- Add a new arity to `make-widget-async` to provide a different widget shape.

## [0.1.1] - 2019-05-18
### Changed
- Documentation on how to make the widgets.

### Removed
- `make-widget-sync` - we're all async, all the time.

### Fixed
- Fixed widget maker to keep working when daylight savings switches over.

## 0.1.0 - 2019-05-18
### Added
- Files from the new template.
- Widget maker public API - `make-widget-sync`.

[Unreleased]: https://github.com/your-name/streamer/compare/0.1.1...HEAD
[0.1.1]: https://github.com/your-name/streamer/compare/0.1.0...0.1.1


streamer/test/streamer/core_test.clj
(ns streamer.core-test
  (:require [clojure.test :refer :all]
            [streamer.core :refer :all]))

(deftest a-test
  (testing "FIXME, I fail."
    (is (= 0 1))))


demo.py
import sys
import time
import json
import math
from kafka import KafkaProducer, KafkaConsumer
import logging
from datetime import datetime, timezone


logging.basicConfig(level=logging.ERROR)
hosts = ['localhost:9092']
topic = 'kafka-demo'
topic_count = 'kafka-demo-count'
topic_derivative = 'kafka-demo-derivative'
ident = "epic-iot-1"

def pretty_datetime(epoch):
    return datetime.fromtimestamp(epoch/1000, timezone.utc).strftime('%Y-%m-%dT%H:%M:%SZ')


def kafka_serializer(data):
	return json.dumps(data).encode('utf-8')


def kafka_deserializer(data):
    return json.loads(data.decode('ascii'))


def kafka_count_deserializer(data):
    return int.from_bytes(data, 'big')


demo_producer = KafkaProducer(bootstrap_servers=hosts, value_serializer=kafka_serializer, request_timeout_ms=1500, max_block_ms=1500)
demo_consumer = KafkaConsumer(topic, bootstrap_servers=hosts, value_deserializer=kafka_deserializer, request_timeout_ms=1500, consumer_timeout_ms=1500)
#demo_count_consumer = KafkaConsumer(topic_count, bootstrap_servers=hosts, value_deserializer=kafka_count_deserializer, request_timeout_ms=1500, consumer_timeout_ms=1500)
demo_derivative_consumer = KafkaConsumer(topic_derivative, bootstrap_servers=hosts, value_deserializer=kafka_deserializer, request_timeout_ms=1500, consumer_timeout_ms=1500)


while True:
    try:
        timestamp_ms = int(round(time.time() * 1000))
        payload = {
            "id": ident,
            "ts": timestamp_ms,
            "value": 1000 * math.sin(((timestamp_ms / 1000) % 360) / 360.0 * 2 * math.pi)
            }
        print(json.dumps(payload))
        demo_producer.send(topic, value=payload, timestamp_ms=timestamp_ms)
        time.sleep(1)

        # consumer all messages every 5 seconds to demonstrate we can read our own messages
        if int(timestamp_ms / 1000.0) % 5 == 0:
            for message in demo_consumer:
                print("Consumer: {ts} {id} {val}".format(ts=pretty_datetime(message.timestamp), id=message.value['id'], val=message.value['value']))
            #for topic_partition, messages in demo_count_consumer.poll().items():
            #    for message in messages:
            #       print("Count consumer: '{ts}' '{val}'".format(ts=pretty_datetime(message.timestamp), val=message.value))
            for topic_partition, messages in demo_derivative_consumer.poll().items():
                for message in messages:
                    #print("Derivative: '{ts}' '{val}'".format(ts=pretty_datetime(message.timestamp), val=message.value))
                    print("Derivative: {ts} {id} {val}".format(ts=pretty_datetime(message.timestamp), id=message.value['id'], val=message.value['difference']))
    except KeyboardInterrupt:
        print("Terminating gracefully...")
        sys.exit()


streamer/project.clj
(defproject streamer "1.0"
  :description "Kafka demo KStream and KTable processor"
  :url "http://github.com/crnkofe"
  :license {:name "Eclipse Public License"
            :url "http://www.eclipse.org/legal/epl-v10.html"}
  :dependencies [[org.clojure/clojure "1.8.0"]
                 [org.clojure/data.json "0.2.6"]
                 [org.apache.kafka/kafka-streams "2.2.0"]
                 [org.apache.kafka/kafka-clients "2.2.0"]
                 [org.apache.kafka/kafka-streams-test-utils "2.2.0"]
                 [org.apache.kafka/kafka-clients "2.2.0"]
                 [org.apache.kafka/connect-json "2.2.0"]
                 [org.springframework.kafka/spring-kafka "2.2.6.RELEASE"]
                 [org.slf4j/slf4j-log4j12 "1.7.1"]
                 [cheshire "5.8.1"]
                 [log4j/log4j "1.2.17" :exclusions [javax.mail/mail
                                                    javax.jms/jms
                                                    com.sun.jmdk/jmxtools
                                                    com.sun.jmx/jmxri]]
                 ]
  :main ^:skip-aot streamer.core
  :target-path "target/%s"
  :profiles {:uberjar {:aot :all}})


streamer/src/streamer/core.clj
(ns streamer.core
  (:import
   (org.apache.kafka.common.serialization Serde Serdes Serializer)
   (org.apache.kafka.streams StreamsConfig KafkaStreams StreamsBuilder)
   (org.apache.kafka.streams KeyValue)
   (org.apache.kafka.common.serialization Serializer)
   (org.apache.kafka.common.serialization Deserializer)
   (org.apache.kafka.streams.kstream ForeachAction)
   (java.io ByteArrayOutputStream)
   (java.io ByteArrayInputStream)
   (org.apache.kafka.streams.kstream ValueMapper)
   (org.apache.kafka.streams.kstream Produced)
   (org.apache.kafka.streams.kstream Serialized)
   (org.apache.kafka.streams.kstream KeyValueMapper)
   (org.apache.kafka.streams.kstream Printed)
   (org.apache.kafka.streams.kstream TimeWindows)
   (org.apache.kafka.streams.kstream Reducer)
   (org.apache.kafka.streams.kstream SessionWindows)
   (org.apache.kafka.streams.kstream WindowedSerdes)
   (org.apache.kafka.streams.kstream Grouped)
   (org.apache.kafka.streams.kstream Materialized)
   (java.time Duration)
)
(:require [cheshire.core :refer :all])
  (:gen-class))

(deftype JsonSerializer [opts]
  Serializer
  (configure [_ _ _])
  (serialize [_ _ data]
    (when data
      (with-open [baos (ByteArrayOutputStream.)
                  writer (clojure.java.io/writer baos)]
        (generate-stream data writer opts)
        (.toByteArray baos))))
  (close [_]))

(defn json-serializer
  "JSON serializer for Apache Kafka.
  Use for serializing Kafka keys values.
  > Notes: You may pass any of the built-in Chesire options for generate-stream via the opts map, using
   the 1-arity version of this function."
  (^JsonSerializer [] (json-serializer nil))
  (^JsonSerializer [opts]
(JsonSerializer. opts)))

(deftype JsonDeserializer [opts]
  Deserializer
  (configure [_ _ _])
  (deserialize [_ _ data]
    (when data
      ;;possible to move this to store in type somewhere
      ;;the parse-stream function doesn't take its options the same way as generate-stream
      ;;we could call apply on it, but want to leave room for future options without breaking signatures
      (let [{:keys [key-fn array-coerce-fn]} opts]
        (with-open [bais (ByteArrayInputStream. data)
                    reader (clojure.java.io/reader bais)]
          (parse-stream reader key-fn array-coerce-fn)))))
  (close [_]))

(defn json-deserializer
  "JSON deserializer for Apache Kafka.
  Use for deserializing Kafka keys and values.
  > Notes: You may pass any of the built-in Cheshire options to parse-stream via the opts map, using
   the 1-arity version of this function."
  (^JsonDeserializer [] (json-deserializer nil))
  (^JsonDeserializer [opts]
(JsonDeserializer. opts)))



(defn kvalue-mapper-default [key payload]
 (let [decoded-payload (decode payload)]
  (KeyValue. (get decoded-payload "ts") {:ts (get decoded-payload "ts"), :id (get decoded-payload "id"), :value (get decoded-payload "value")})
 ))

;; attach seconds timestamp as key to stream
(defn kvalue-mapper [key payload]
 (let [decoded-payload (decode payload)]
  (KeyValue. (quot (get decoded-payload "ts") 1000) {:ts (get decoded-payload "ts"), :id (get decoded-payload "id"), :value (get decoded-payload "value")})
 ))


(defn kvalue-mapper-seconds [key payload]
 (let [decoded-payload (decode payload)]
  (KeyValue. (quot (get decoded-payload "ts") 1000) {:ts (get decoded-payload "ts"), :id (get decoded-payload "id"), :value (get decoded-payload "value")})
 ))


(defn kvalue-mapper-minutes [key payload]
 (let [decoded-payload (decode payload)]
  (KeyValue. (quot (quot (get decoded-payload "ts") 1000) 60) {:ts (get decoded-payload "ts"), :id (get decoded-payload "id"), :value (get decoded-payload "value")})
 ))


(defn kvalue-mapper-minutes-five [key payload]
 (let [decoded-payload (decode payload)]
  (KeyValue. (quot (quot (quot (get decoded-payload "ts") 1000) 60) 5) {:ts (get decoded-payload "ts"), :id (get decoded-payload "id"), :value (get decoded-payload "value")})
 ))

(defn identity-map [input-topic]
  (let [builder (StreamsBuilder.)]
    (->
     (.stream builder input-topic) ;; Create the source node of the stream
     (.map (reify KeyValueMapper (apply [_ k v] (kvalue-mapper k v))))
	 (.print (Printed/toSysOut )))
    builder))

(defn identity-stream [input-topic]
  (let [builder (StreamsBuilder.)
        window (.advanceBy (TimeWindows/of  60000) 1000)
		jsonSerdes (Serdes/serdeFrom (JsonSerializer. []) (JsonDeserializer. []))
		serializer (Serialized/with (Serdes/String) (Serdes/Long))
		string-serializer (Serialized/with (Serdes/String) (Serdes/Long))]
	(try
	  (->
		  (.stream builder input-topic) ;; Create the source node of the stream
		  (.map (reify KeyValueMapper (apply [_ k v] (kvalue-mapper-minutes k v))))
		  (.to "kafka-demo-count" (Produced/with (Serdes/Long) jsonSerdes)))
      (catch Exception e (println (str "caught exception: " (.toString e))))
      (finally (println "This is our final block")))
    builder))

(defn count-stream [input-topic]
  (let [builder (StreamsBuilder.)
        window (.advanceBy (TimeWindows/of  60000) 1000)
		jsonSerdes (Serdes/serdeFrom (JsonSerializer. []) (JsonDeserializer. []))
		serializer (Serialized/with (Serdes/String) (Serdes/Long))
		string-serializer (Serialized/with (Serdes/String) (Serdes/Long))]
	(try
	  (->
		  (.stream builder input-topic) ;; Create the source node of the stream
		  (.map (reify KeyValueMapper (apply [_ k v] (kvalue-mapper-minutes k v))))
          (.groupByKey (Serialized/with (Serdes/Long) jsonSerdes))
          (.count)
          (.toStream)
		  (.to "kafka-demo-count" (Produced/with (Serdes/Long) jsonSerdes)))
      (catch Exception e (println (str "caught exception: " (.toString e))))
      (finally (println "This is our final block")))
    builder))

(defn count-window-stream [input-topic]
  "Group timestamped values by 5 minutes
  Window them by 1 minute non-overlapping intervals
  and compute counted values in that window
  output that to a different topic
  "
  (let [builder (StreamsBuilder.)
        window (.advanceBy (TimeWindows/of  60000) 1000)
		jsonSerdes (Serdes/serdeFrom (JsonSerializer. []) (JsonDeserializer. []))
        windowSerdes (WindowedSerdes/timeWindowedSerdeFrom Long)
		serializer (Serialized/with (Serdes/String) (Serdes/Long))
		string-serializer (Serialized/with (Serdes/String) (Serdes/Long))]
	(try
	  (->
		  (.stream builder input-topic) ;; Create the source node of the stream
		  (.map (reify KeyValueMapper (apply [_ k v] (kvalue-mapper-minutes-five k v))))
          (.groupByKey (Serialized/with (Serdes/Long) jsonSerdes))
          (.windowedBy (.advanceBy (TimeWindows/of (Duration/ofMinutes 1) ) (Duration/ofSeconds 60)))
          (.count (Materialized/with (Serdes/Long) (Serdes/Long)))
          (.toStream)
		  (.to "kafka-demo-count" (Produced/with windowSerdes (Serdes/Long))))
      (catch Exception e (println (str "caught exception: " (.toString e)))))
    builder))

(defn difference [p n]
  (let [diff (- (get n "value") (get p "value"))]
    ;; (println diff (get p "ts") (get n "ts"))
    ;; (println diff (get p "value") (get n "value"))
    {"ts" (get n "ts"), "id" (get n "id"), "value" (get n "value"), "difference" diff}
  )
  )

(defn derivative-stream [input-topic]
  "Group timestamped values by 1 minute

  and compute counted values in that window
  output that to a different topic
  "
  (let [builder (StreamsBuilder.)
        window (.advanceBy (TimeWindows/of  60000) 1000)
		jsonSerdes (Serdes/serdeFrom (JsonSerializer. []) (JsonDeserializer. []))
        windowSerdes (WindowedSerdes/timeWindowedSerdeFrom Long)
		serializer (Serialized/with (Serdes/String) (Serdes/Long))
		string-serializer (Serialized/with (Serdes/String) (Serdes/Long))]
	(try
	  (->
		  (.stream builder input-topic) ;; Create the source node of the stream
		  (.map (reify KeyValueMapper (apply [_ k v] (kvalue-mapper-minutes k v))))
          (.groupByKey (Serialized/with (Serdes/Long) jsonSerdes))
          (.windowedBy (.advanceBy (TimeWindows/of (Duration/ofSeconds 10) ) (Duration/ofSeconds 10)))
          (.reduce (reify Reducer (apply [_ p n] (difference p n))))
          (.toStream)
		  (.to "kafka-demo-derivative" (Produced/with windowSerdes jsonSerdes)))
      (catch Exception e (println (str "caught exception: " (.toString e)))))
    builder))

(defn -main [& args]
  (def properties
    {StreamsConfig/APPLICATION_ID_CONFIG, "streamer"
     StreamsConfig/BOOTSTRAP_SERVERS_CONFIG, "localhost:9092"
     StreamsConfig/DEFAULT_KEY_SERDE_CLASS_CONFIG, (.getName (.getClass (Serdes/Long)))
     StreamsConfig/DEFAULT_VALUE_SERDE_CLASS_CONFIG, (.getName (.getClass (Serdes/String)))})

  (def config
    (StreamsConfig. properties))

  (def input-topic "kafka-demo")

  (def streams
    (KafkaStreams. 
      (.build (derivative-stream input-topic)) config))

  (.start streams)
  (Thread/sleep (* 60000 10))
)
